CacheSim/CacheSimLinux.cpp

```diff
@@ -0,0 +1,575 @@
+  if ( modules->m_Count == ARRAY_SIZE(modules->m_Infos) )
+  {
+    fprintf(stderr, "Cannot record additional modules. Increase the ModuleList::m_Infos buffer size.\n");
+    return -1;
+  }
+
+  for ( ElfW(Half) i = 0; i < info->dlpi_phnum; i++ )
+  {
+    const ElfW(Phdr)& header = info->dlpi_phdr[i];
+
+    // Find the executable segment
+    if ( (header.p_flags & PF_X) && (header.p_type == PT_LOAD) )
+    {
+      ModuleInfo& module = modules->m_Infos[modules->m_Count];
+      if ( info->dlpi_addr == getauxval(AT_SYSINFO_EHDR) )
+      {
+        // vdso section.
+        continue;
+      }
+      
+      module.m_StartAddrInMemory = (void*)info->dlpi_addr;
+      module.m_SegmentOffset = (void*)header.p_vaddr;
+      module.m_Length = header.p_memsz;
+
+      bool hasName = (info->dlpi_name != nullptr) && (info->dlpi_name[0] != '\0');
+      if ( !hasName && (modules->m_ModuleCallbacks == 1) ) 
+      {
+        // This is the main executable.
+        module.m_Filename = executable_filename;
+      }
+      else if (hasName)
+      {
+        module.m_Filename = info->dlpi_name; // dlpi_name will continue to exist--at least on this implementation of linux :)
+      }
+      else
+      {
+        fprintf(stderr, "Failed to get name for module at address!!: %p\nSkipping\n", (int*)info->dlpi_addr);
+        continue;
+      }
+      modules->m_Count++;
+      return 0;
+    }
+  }
+
+  return 0;
+}
+
+namespace
+{
+  template<typename T> void WriteHelper(FILE* f, const T& val)
+  {
+    fwrite(&val, 1, sizeof val, f);
+  }
+}
+
+void CacheSimEndCapture(bool save)
+{
+  using namespace CacheSim;
+  g_TraceEnabled = 0;
+
+  // Clear our trap flag
+  asm volatile("pushf\n"
+    "andl $0xFFFFFFFFFFFFFEFF, (%rsp)\n"
+    "popf\n");
+
+
+  // Give the thread a few instructions to run so we're definitely not tracing.
+  //Thread::NanoPause();
+  usleep(0);
+  usleep(0);
+  usleep(0);
+
+  if ( !save )
+    return;
+
+  AutoSpinLock lock;
+
+  // It's tempting to remove the signal handler here
+  //
+  //    RemoveVectoredExceptionHandler(g_Handler);
+  //    g_Handler = nullptr;
+  //
+  // ..but that's a mistake. There could be a syscall instruction paused in the kernel that
+  // will come back and signal a single step trap at some arbitrary point in the future, so
+  // we need our handler to stay in effect.
+
+  const char* executable_name = "unknown";
+  if ( char* p = strrchr(executable_filename, '/') )
+  {
+    ++p;
+    executable_name = p;
+  }
+
+  char fn[512];
+  snprintf(fn, ARRAY_SIZE(fn), "%s_%u.csim", executable_name, (uint32_t)time(nullptr));
+
+  if ( FILE* f = fopen(fn, "w") )
+  {
+    auto align = [&f]()
+    {
+      if ( int needed = (8 - (ftell(f) & 7)) & 7 )
+      {
+        static const uint8_t padding[8] = { 0 };
+        fwrite(padding, 1, needed, f);
+      }
+    };
+
+#define welem(value) WriteHelper(f, value)
+
+#define wdata(data, size) fwrite(data, 1, size, f);
+
+    struct PatchWord
```

Comment:
Probably. It'll take some refactoring to pull out the EnumModules and related code into something like the dl_iterate_phdr version, but it shouldn't be hard. The bigger issue is that the version of GCC I'm stuck on doesn't support C++14, so I can't use the lambdas the way you did. So I can either use the gcc #define version everywhere or provide alternative implementations guarded by "#if linux". I'll take a look tomorrow and see if there are better options.